File: solutions/cpp/Count_occurrences_in_sentence.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <iostream>
#include <random>
#include <string>
#include <unordered_map>
#include <cassert>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::unordered_map;

static unordered_map<char, int> charToCount;

// @include
void CountOccurrences(string S) {
  sort(S.begin(), S.end());

  int current_character_count = 1;
  for (int i = 1; i < S.size(); ++i) {
    if (S[i] == S[i - 1]) {
      ++current_character_count;
    } else {
      cout << '(' << S[i - 1] << ',' << current_character_count << "),";
      // @exclude
      charToCount[S[i-1]] = current_character_count;
      // @include
      current_character_count = 1;
    }
  }
  cout << '(' << S.back() << ',' << current_character_count << ')' << endl;
}
// @exclude

string RandString(int len) {
  default_random_engine gen((random_device())());
  string ret;
  while (len--) {
    uniform_int_distribution<int> dis('a', 'z');
    ret += dis(gen);
  }
  return ret;
}

static void SimpleTest() {
  CountOccurrences("foo bar! ABA A");
  assert(charToCount['f'] == 1);
  assert(charToCount.count('F') == 0);
  assert(charToCount['o'] == 2);
  assert(charToCount.count('x') == 0);
  assert(charToCount[' '] == 3);
  assert(charToCount['!'] == 1);
  assert(charToCount['A'] == 3);
  assert(charToCount['B'] == 1);
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  string S;
  if (argc == 2) {
    S = argv[1];
  } else {
    uniform_int_distribution<int> dis(1, 1000);
    S = RandString(dis(gen));
  }
  cout << S << endl;
  CountOccurrences(S);
  return 0;
}

File: static/solutions/cpp/java/src/main/java/com/epi/CountOccurrencesInSentence.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class CountOccurrencesInSentence {
  private static Map<Character, Integer> charToCount = new HashMap<>();

  // @include
  public static void countOccurrences(String s) {
    char[] a = s.toCharArray();
    Arrays.sort(a);

    int currentCharacterCount = 1;
    for (int i = 1; i < a.length; ++i) {
      if (a[i] == a[i - 1]) {
        ++currentCharacterCount;
      } else {
        System.out.print("(" + a[i - 1] + "," + currentCharacterCount + "),");
        // @exclude
        charToCount.put(a[i - 1], currentCharacterCount);
        // @include
        currentCharacterCount = 1;
      }
    }
    System.out.println("(" + a[a.length - 1] + ',' + currentCharacterCount
                       + ")");
  }
  // @exclude

  private static String randomString(int len) {
    StringBuilder ret = new StringBuilder();
    Random rnd = new Random();

    while (len-- > 0) {
      ret.append((char)(rnd.nextInt(26) + 97));
    }
    return ret.toString();
  }

  private static void simpleTest() {
    countOccurrences("foo bar! ABA A");
    assert(charToCount.get('f') == 1);
    assert(charToCount.get('F') == null);
    assert(charToCount.get('o') == 2);
    assert(charToCount.get('x') == null);
    assert(charToCount.get(' ') == 3);
    assert(charToCount.get('!') == 1);
    assert(charToCount.get('A') == 3);
    assert(charToCount.get('B') == 1);
  }

  public static void main(String[] args) {
    simpleTest();
    Random rnd = new Random();
    String s;
    if (args.length == 1) {
      s = args[0];
    } else {
      s = randomString(rnd.nextInt(1000) + 1);
    }
    System.out.println(s);
    countOccurrences(s);
  }
}
